transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -676,27 +633,29 @@ public ChannelFuture shutdown() {
 
     @Override
     public ChannelFuture shutdown(final ChannelPromise promise) {
-        Executor closeExecutor = ((EpollStreamUnsafe) unsafe()).prepareToClose();
-        if (closeExecutor != null) {
-            closeExecutor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    shutdown0(promise);
-                }
-            });
-        } else {
-            EventLoop loop = eventLoop();
-            if (loop.inEventLoop()) {
-                shutdown0(promise);
-            } else {
-                loop.execute(new Runnable() {
+        shutdownOutput().addListener(new ChannelFutureListener() {
```

Comment:
if we are on the event loop and so_linger isn't set this will complete synchronously. consider adding the typical check for if the future is complete before adding a `ChannelFutureListener` in both these cases.

Reply:
yep let me fix and also for other transports.